### ISSUE TITLE
How to initialize endpoint and connect streams (I210, open-avb)

### ISSUE THREAD

**Issue by owenfmurphy** [audioscience/avdecc-lib#323]
Hello,
Is there any tutorial for using the command line program to initialize the machine as an endpoint (acquire entity number, etc)?  Currently I have an Ubuntu PC with an I210 PCI card, and I am trying to get it to connect to my AVB network. I have enabled the open-avb drivers, and I am running the open-avb daemons.  The rest of the network consists of:
(1) Mac Mini (latest OS, AVB compatible ethernet chip)
(1) MOTU AVB interface
(1) DSP4You OEM AVB endpoint
(1) MOTU AVB Switch

When I run the command line program, I can see the Mac Mini, the Motu interface, and the endpoint when I type "view all" and "list", but the ubuntu machine does not appear.  Is there an command I need to issue that allows the ubuntu machine to connect to the network?

There is another strange issue I noticed regarding the mac mini. When I type "view all", none of the parameters for the mac mini show up the way they do for the other endpoints.  The streams are available though, because I can connected them to the motu interface using motu's browser based controller.  Does anyone know if OS X follows the AVDECC protocol exactly?  I am having problems using my mac mini with the switch, so it's possible they don't implement it the protocol properly.  I just wanted to see if anyone else had issues, or has any information on the matter.

Thanks for the help!
Owen

**Comment by jdkoftinoff** [comment]
At this point the OpenAVB code does not include an AVDECC entity implementation.  The starting point for writing one is my 1722.1 reference C code; I had started an entity model implementation for constrained devices at https://github.com/jdkoftinoff/jdksavdecc-mcu <https://github.com/jdkoftinoff/jdksavdecc-mcu> but it is incomplete.

macOS Sierra includes full compliant AVDECC protocol implementation.  The default mode will allow you to see compatible AVB Audio devices in the “Audio MIDI Setup” application in the “Network browser” window. But there is more powerful options with the “avbutil” command line tool.

To see how to create a virtual AVB audio device and manage streams, run:

man avbutil

Specifically the “ --virtual-audio enable” and “--controller launch” options are useful!

Jeff

**Comment by owenfmurphy** [comment]
Thanks for the quick reply, Jeff!

Since Open-AVB does not include entity implementation, does that mean you can only use the prewritten examples with with each other ("simpler_talker" <-> "simple_listener") and not with a standard AVB device ("simple talker" <-> off the shelf AVB interface)?

Thanks for the info on avbutil in OS X.  Do you have any idea why the parameters for my Mac don't appear when I use the command "view all" in the avdecccmdline program? Virtual audio entity is enabled, and there is AVB audio data being transferred between the Mac and the interface.  But for some reason, I can't see any of the parameters.

When I type "view all", this is what I get this for the DSP4YOU endpoint and MOTU interface:

`End Station: 0 (DSP4YOU AVB Dongle)
     Descriptor Type   Descriptor Index        Descriptor Name
------------------------------------------------------------------------------`
`ENTITY                  0     DSP4YOU AVB Dongle
       CONFIGURATION                  0        Configuration 0

Top Level Descriptors
          AUDIO_UNIT                  0           Audio Unit 0
        STREAM_INPUT                  0                Input 0`
etc etc

Only this shows up for the Mac: 

`End Station: 2 (Owen’s Mac mini)
     Descriptor Type   Descriptor Index        Descriptor Name
------------------------------------------------------------------------------`

and nothing else.

I have two several other AVDECC programs (Pivitec AVDECC Controller and AudioScience Hono Controller) and both of them also don't show available streams from the Mac.


I have another issue that is unrelated to this software, but I thought maybe you'd have some insight.  When I connect my interface directly to the Mac, I am able to use three AVB streams in each direction (the maximum allowed by the interface).  When I put a switch in between the Mac and the interface, I can only use one stream.  I don't think the switch is the issue, since I've used both a MOTU switch and a DSP4YOU switch, both of which are designed for handling AVB traffic.  I talked to the interface manufacturer and they basically said it was on Apple's end, but didn't really clarify.  This issue is actually the whole reason I started looking into these open source AVB resources, hoping to figure out the problem.  Do you have any idea why that could be?  The interface AVB connections on the interface appear to be locked so I can't use this utility to manually connect streams.

Thanks again,
Owen

**Comment by jdkoftinoff** [comment]
How did you connect the streams between the virtual audio entity and the interface?

As for your secondary issue:  What are the dimensions of the 3 streams? 48 kHz 8 channel AM824?

Jeff

**Comment by owenfmurphy** [comment]
The interface has a web browser based controller that lets you connect streams with pulldown menus.  The virtual audio entity streams connected automatically to the interface streams.  (In fact they were locked, so I was unable to disconnect them...)


The three streams are 8 channels each at 48kHz, 24-bit depth

**Comment by owenfmurphy** [comment]
This is the browser based controller

![screen shot 2016-12-27 at 10 09 05 pm](https://cloud.githubusercontent.com/assets/24628294/21512999/2b81bcd8-cc81-11e6-8b3c-6e28e83b879b.png)

**Comment by jdkoftinoff** [comment]
I don’t think you actually used the macOS virtual entity; When the mac sees a directly connected AVB interface (without a switch in between) the built in controller in macOS tries to automatically connect all the streams. Run “Audio MIDI Setup” and open the “Network Device Browser” window to see.

If you were using your macOS virtual audio entity then you would have had to connect the streams manually via a Controller; either the avdecccmdline tool running on a different computer, or the built in macOS GUI AVB Controller via the cmdline "avbutil --controller launch"

Regards,
Jeff

**Comment by owenfmurphy** [comment]
Everything looks like it's enabled with the virtual entity.  Is there something else to check?
![screen shot 2016-12-27 at 10 16 18 pm](https://cloud.githubusercontent.com/assets/24628294/21513064/2649c6ec-cc82-11e6-8894-1596931b50c1.png)

**Comment by jdkoftinoff** [comment]
No, that is not the virtual entity as created by “avbutil --virtual-audio enable [interface_name]” What you are seeing is the mirror entity that is auto-generated by the mac to match the MOTU device. This device is not controllable by an external controller.

What does “avbutil --virtual-audio list” report?

And what does “avbutil --controller launch” show?

Jeff

**Comment by owenfmurphy** [comment]
Owens-Mac-mini:~ owenfmurphy$ avbutil --virtual-audio list
avbutil 500.19

Virtual Audio Entities enabled on interface(s):
    en0

Owens-Mac-mini:~ owenfmurphy$ avbutil --controller launch
avbutil 500.19
![screen shot 2016-12-27 at 10 28 46 pm](https://cloud.githubusercontent.com/assets/24628294/21513203/ddd11800-cc83-11e6-8c80-92443181cae4.png)

**Comment by jdkoftinoff** [comment]
Ah, ok! I see.  At this point with the mac connected to the switch you should be able to enumerate the  “Owen’s Mac Mini” entity with the audioscience command line tool.  

Can you try disconnect the streams via the web interface and try connect them via the audioscience command line tool, or try use the avbutil controller GUI to connect.

Jeff

**Comment by owenfmurphy** [comment]
The web interface actually won't let me disconnect the streams.  They are grayed out and I can't click on them.  

When I used another graphical AVDECC tool (AudioScience Hono Controller I believe), it told me the MOTU device was locked, and would not let me unlock it.  Is there a way to unlock an AVB device with the command line tool?

**Comment by jdkoftinoff** [comment]
How exactly did you create the virtual entity?

**Comment by owenfmurphy** [comment]
It happens automatically.  I have a MOTU "discovery app" on the machine that tells you when there is a MOTU device connected.  It's possible that the discovery app creates the virtual entity when it detects a MOTU device.

**Comment by jdkoftinoff** [comment]
Right, so this is not the virtual entity that can be controlled by an external controller.

First step is to use a switch. Second step is to de-select MOTU from the Audio MIDI setup window (to un-acquire it).  Third step is to create a virtual entity manually with the avbutil command. Fourth step is to use an external controller (either the avbutil controller running locally on the Mac or avdecccmdline tool running on a separate linux box) to enumerate the entities and connect the streams.

Jeff

**Comment by owenfmurphy** [comment]
Ok great! I performed steps 1-3 and managed to disconnect the all the streams.  How do I enumerate the entities and connect the streams using avbutil and/or avdecccmdline (I have the ability to do both)

**Comment by jdkoftinoff** [comment]
Best to ask the audioscience team about the latest usage of avdecccmdline. It was pretty straightforward last time I ran it; I selected the network port, it discovered the entities, and I was able to view stream info and connect streams.

As for avbutil, run the controller with
	avbutil --controller launch

and go into the window menu and choose “AVDECC Connection Matrix” and connect your streams.

**Comment by owenfmurphy** [comment]
I was able to streams using the AVDECC Connection Matrix, but each connection is stuck at "Connected, not acquired" (even after 15 minutes)
![screen shot 2016-12-27 at 11 51 35 pm](https://cloud.githubusercontent.com/assets/24628294/21514142/afa5d2c0-cc8f-11e6-9857-a43a7d1583b1.png)

This might be the macOS issue MOTU was talking about.  It was supposed to be fixed with OS 10.12, but clearly something still isn't working.  I'm going to follow up with MOTU tech support to see if they have an explanation.

I'm still unclear as to why avbutil can see the Mac AVB streams ("Owen's Mac Mini: Audio Input Stream x" in the AVDECC Connection Matrix), but I can't see them using any of the other AVDECC tools.

Anyway, thank you so much for the help.  I really do appreciate it.

Owen

**Comment by jdkoftinoff** [comment]
"Connected, not acquired” is reasonable mode for the connections. What do you mean by “stuck”? i.e. are you unable to disconnect?
 
Are the other tools running on a different computer?

**Comment by owenfmurphy** [comment]
By stuck, I meant it stays in "Connected, not acquired".  For some reason I thought that was an intermediate state between its disconnected state, and its transmitting state.  I guess that's not the case, so everything in that regard is OK.

OK, now the original problem is back.  All the connections say "Connected, not acquired", but only one of the three streams actually transmits audio.  This was the same issue I was having originally when the Mac was connected with its non-entity mode.  It seems like the presence of a switch causes this issue.


The other tools are both on this computer and on other computers on the network.  I can't see the Mac's AVB streams In neither case.

**Comment by jdkoftinoff** [comment]
“Not acquired” means that any other controller can disconnect the streams or change settings on the entities. 

Does the working stream depend on the order that you try to connect them?

Jeff

**Comment by owenfmurphy** [comment]
Yes, it does depend on the order.  The working stream is the first stream I try to connect.

**Comment by andrew-elder** [comment]
Are you able to retest with an Extreme AVB switch? What you are trying to do should work in my opinion.

**Comment by vronchetti** [comment]
Hi Owen,

Thank you for reporting these issues.

>There is another strange issue I noticed regarding the mac mini. When I type "view all", none of the parameters for the mac mini show up the way they do for the other endpoints. 

This has been fixed in avdecc-lib but has not made its way from Staging to Master.  I'm assuming you are building off of the Master branch?

>How do I enumerate the entities and connect the streams using avbutil and/or avdecccmdline (I have the ability to do both)



<img width="588" alt="screen shot 2017-01-03 at 10 16 01 am" src="https://cloud.githubusercontent.com/assets/8479744/21612091/b3740fbc-d19d-11e6-926e-37b12a673b02.png">


>I'm still unclear as to why avbutil can see the Mac AVB streams ("Owen's Mac Mini: Audio Input Stream x" in the AVDECC Connection Matrix), but I can't see them using any of the other AVDECC tools.

We will shortly be releasing a new version of AudioScience Hono AVB Controller that will support macOS virtual entities and other multi-configuration AVB devices.  Until then you will not be able to see Mac streams.

>OK, now the original problem is back. All the connections say "Connected, not acquired", but only one of the three streams actually transmits audio.

Could you provide a screenshot of what you see in MOTU web interface Device and Routing pages when you are in this state?

Regards,

Victor

**Comment by owenfmurphy** [comment]
Andrew:
I have tried using a MOTU AVB switch and a DSP4YOU switch, which should both be compliant (right?). I have the same issue with both switches.
http://motu.com/products/avb/avb-switch
http://www.dsp4you.com/products/avb-oem-series/avb-sw
I don't have access to an extreme networks switch, and since this is a hobby project, I probably won't purchase one unless I have to.  They are over $1000 each, right?  

Have their been issues with switches other than Extreme Networks swtiches in the past?


Victor:
Thanks for the info.  To answer your question, I am building from the master branch.

Here's a screenshot of the MOTU web interface:
![screen shot 2017-01-03 at 8 21 44 pm](https://cloud.githubusercontent.com/assets/24628294/21629062/c7bb97ba-d1f2-11e6-924e-38e657787a8c.png)
The label next to "Input Stream 1" is solid and the labels next to "Input Stream 2" and "Input Stream 3" are blinking.  Input stream 1 is the only stream that actually has audio coming out of it.

Here's a screenshot of the avbutil routing matrix 
![screen shot 2017-01-03 at 8 26 18 pm](https://cloud.githubusercontent.com/assets/24628294/21629150/700015fe-d1f3-11e6-8bd4-e8663006ecd8.png)

Thanks for the help
Owen

**Comment by andrew-elder** [comment]
Owen,
Regarding switches, the Extreme switch is the only one on the market that has passed Avnu's certification testing and has been proven to work in large AVB installations. That said, I unaware of any particular issues with the MOTU switch. If you had an Extreme switch available for testing, that would be the next logical test to perform. Have you tried contacting MOTU about this?
- Andrew

**Comment by jdkoftinoff** [comment]
FYI my understanding is that the motu ultralight has a 100baseT Ethernet port.   I wonder if that is relevant here.

Perhaps try update the firmware in your motu switch?

**Comment by owenfmurphy** [comment]
I've been in contact with MOTU and they essentially told me that it was an issue with the apple OS and that apple developers were aware of the issue. It was supposed to have been fixed with the 10.12 Sierra release, but for whatever reason, it still does not work.   I've let MOTU know that it still doesn't work after the OS upgrade, but I haven't gotten any more info from them.

Has anyone been able to successfully use a switch (any brand) with multiple AVB streams with OS X?

The MOTU switch and the DSP4YOU switch both have the most current firmware.  Both switches use Marvell chips - probably not relevant though.

**Comment by jdkoftinoff** [comment]
Thanks for the info.

**Comment by andrew-elder** [comment]
We have streamed 4 streams of 8 channels each to an ASI2620 (unreleased firmware) through an Extreme switch. We did not observe any issues with that configuration. The ASI2620 uses a 1G NIC.

**Comment by rockermatt** [comment]
Hi all,
sorry, very basic question/issue:
I've connected 2 macs as virtual audio entities, daisy chained.  They see each other in audio/midi setup, and I'm able to pass audio from one Pro Tools system to another.  (8 streams x 8 channels)
I do not have the "AVDECC Connection Matrix" menu item in the "window" menu of the AVB Audio Configuration app however.
Is that something that should be there by default, or that needs to be enabled via command line?
Or do I need to use a compliant switch to have that option?  (I'm directly connected from Mac A's ethernet port to Mac B's ethernet port, no switch, hub, or router)
thanks,
matt
